### PR TITLE
293504 accredidation limit 2

### PR DESCRIPTION
**Pull request by AtosSaqlain**

### CODE REVIEW

File: EPRN.Common/Constants/Constants.cs [hunk 1]
```diff
@@ -102,7 +102,14 @@ public static class Waste
...
                 public const string DecemberWaste = "DecemberWaste";
+                public const string AccredidationLimit = "AccredidationLimit";
             }
         }
     }
+}
+
+public static class Double
+{
+    public static readonly double AccredidationLimit = 400;
```
Comment (NikPeters-Atos) [inline EPRN.Common/Constants/Constants.cs]: Should this not be configurable?

File: EPRN.Portal/Controllers/WasteController.cs [hunk 2]
```diff
@@ -20,16 +20,20 @@ public class WasteController : BaseController
     {
         private readonly IWasteService _wasteService;
         private readonly IUserBasedService _homeService;
+        private readonly WasteCommonViewModel _wasteCommonViewModel;
 
         public WasteController(
             IWasteService wasteService,
-            IHomeServiceFactory homeServiceFactory)
+            IHomeServiceFactory homeServiceFactory, WasteCommonViewModel wasteCommonViewModel)
```
Comment (NikPeters-Atos) [inline EPRN.Portal/Controllers/WasteController.cs]: Multiple parameters read better on individual lines

File: EPRN.Portal/Controllers/WasteController.cs [hunk 3]
```diff
@@ -199,20 +204,43 @@ public async Task<IActionResult> Tonnes(int? id)
...
+
+            if (performLimitCheck && exportTonnageViewModel.ExportTonnes.HasValue)
+            {
+                var accredidationLimitViewModel = await _wasteService.GetAccredidationLimit(
+                    exportTonnageViewModel.Id, 
+                    _wasteCommonViewModel.CompanyReferenceId, 
+                    exportTonnageViewModel.ExportTonnes.Value);
+
+                if (accredidationLimitViewModel.ExcessOfLimit < 0)
+                    return View("AccredidationLimit", accredidationLimitViewModel);
```
Comment (NikPeters-Atos) [inline EPRN.Portal/Controllers/WasteController.cs]: Add the name of the view to the Routes constants class

File: EPRN.Portal/Controllers/WasteController.cs [hunk 4]
```diff
@@ -199,20 +204,43 @@ public async Task<IActionResult> Tonnes(int? id)
...
             return RedirectToAction(
                 Routes.Actions.Waste.Baled, 
                 new { id = exportTonnageViewModel.Id });
         }
 
+        [HttpPost]
+        [ActionName(Routes.Actions.Waste.AccredidationLimit)]
+        public IActionResult AccredidationLimitPost(int? id)
+        {
+            return RedirectToAction("Tonnes", "Waste", new { id });
```
Comment (NikPeters-Atos) [inline EPRN.Portal/Controllers/WasteController.cs]: Doesn't seem to do anything, should it?

Comment (NikPeters-Atos) [inline EPRN.Portal/Controllers/WasteController.cs]: Use Routes constant class for Action name and Controller name

File: EPRN.Portal/Services/WasteService.cs [hunk 5]
```diff
@@ -394,5 +395,23 @@ public async Task<string> GetWasteType(int journeyId)
         {
             return await _httpJourneyService.GetWasteType(journeyId);
         }
+
+        public async Task<AccredidationLimitViewModel> GetAccredidationLimit(int journeyId, string userReferenceId, double newQuantityEntered)
+        {
+            var accredidationLimitDto = await _httpJourneyService.GetAccredidationLimit(journeyId, userReferenceId, newQuantityEntered);
+
+            var vm = new AccredidationLimitViewModel
```
Comment (NikPeters-Atos) [inline EPRN.Portal/Services/WasteService.cs]: Use Automapper here

File: EPRN.Portal/Views/Waste/AccredidationLimit.cshtml [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+@{
+    ViewData["Title"] = @AccredidationLimitResources.Title;
+}
+
+<partial name="_ValidationSummary" />
+<partial name="_WasteTypeName" />
+
+<form method="post" action="Tonnes">
+    <input type="hidden" name="Id" value="@Model.JourneyId" />
```
Comment (NikPeters-Atos) [inline EPRN.Portal/Views/Waste/AccredidationLimit.cshtml]: Don't need this - ID is provided through the routes binding

File: EPRN.Portal/Views/Waste/AccredidationLimit.cshtml [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+                    <p class="govuk-hint">
+                        @AccredidationLimitResources.Line1_Part1
+                        <a href="#">
+                            @AccredidationLimitResources.Line1_Part2
+                        </a>     
+                        @AccredidationLimitResources.Line1_Part3
+                    </p>
+                    <p class="govuk-hint">@AccredidationLimitResources.Line2.Replace("<<newTonnageEntered>>", Model.NewAmountEntered.ToString()).Replace("<<exceedLimitBy>>", Model.ExcessOfLimit.ToString())</p>
+                    <p class="govuk-hint">@AccredidationLimitResources.Line3.Replace("<<currentAccredidationLimit>>", Model.AccredidationLimit.ToString())</p>
+                    <p class="govuk-hint">@AccredidationLimitResources.Line4.Replace("<<totalPrnTonnageToDate>>", Model.TotalToDate.ToString())</p>
```
Comment (NikPeters-Atos) [inline EPRN.Portal/Views/Waste/AccredidationLimit.cshtml]: Could have used standard string formatting replacement here 
(i.e. string.Format("My string. Parameter1: {0}, Parameter2: {1}", param1, param2);)

File: EPRN.UnitTests/Portal/Controllers/WasteControllerTests.cs [hunk 8]
```diff
@@ -583,5 +611,64 @@ public async Task CheckYourAnswers_ReturnsRedirectToAction_WhenModelIsValid()
             Assert.AreEqual("Index", redirectResult.ActionName);
             Assert.AreEqual("Home", redirectResult.ControllerName);
         }
+
```
Comment (NikPeters-Atos) [inline EPRN.UnitTests/Portal/Controllers/WasteControllerTests.cs:614]: Either remove commented out code completely or uncomment

File: EPRN.Common/Constants/Constants.cs (hunk 1 above)
Comment (AtosSaqlain) [reply EPRN.Common/Constants/Constants.cs]: According to Kate, this is a fixed figure and will not change. If/when that situation changes we can move it to a config file.

File: EPRN.Portal/Views/Waste/AccredidationLimit.cshtml (hunk 6 above)
Comment (AtosSaqlain) [reply EPRN.Portal/Views/Waste/AccredidationLimit.cshtml]: The ID provided by route binding is consumed on the first pass-through the [Tonnes] action. 

On the second pass through (if needed) the ID is passed via hidden field in the view which is then "re-bound" to the route.

File: EPRN.Common/Constants/Constants.cs (hunk 1 above)
Comment (NikPeters-Atos) [reply EPRN.Common/Constants/Constants.cs]: I would still recomend having it configurable. If it doesn't change, then there's no issue, but if it ever does, then it's an easy change.

File: EPRN.Portal/Views/Waste/AccredidationLimit.cshtml (hunk 6 above)
Comment (NikPeters-Atos) [reply EPRN.Portal/Views/Waste/AccredidationLimit.cshtml]: I don't understand what you mean.

**Review by NikPeters-Atos — APPROVED** [review]

**Pull request merged**